src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java

```diff
@@ -0,0 +1,83 @@
+package com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeSmell;
+
+/*
+ * This file is part of the Php Inspections (EA Extended) package.
+ *
+ * (c) Funivan <alotofall@gmail.com>
+ * (c) Vladimir Reznichenko <kalessil@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+import com.intellij.codeInspection.ProblemHighlightType;
+import com.intellij.codeInspection.ProblemsHolder;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiElementVisitor;
+import com.intellij.psi.PsiRecursiveElementWalkingVisitor;
+import com.jetbrains.php.lang.psi.elements.Function;
+import com.jetbrains.php.lang.psi.elements.Method;
+import com.jetbrains.php.lang.psi.elements.PhpClass;
+import com.jetbrains.php.lang.psi.elements.PhpReturn;
+import com.jetbrains.php.lang.psi.visitors.PhpElementVisitor;
+import com.kalessil.phpStorm.phpInspectionsEA.openApi.BasePhpInspection;
+import com.kalessil.phpStorm.phpInspectionsEA.utils.ExpressionSemanticUtil;
+import org.jetbrains.annotations.NotNull;
+
+public class MultipleReturnPointsInspector extends BasePhpInspection {
+    private static final String message = "Method with multiple return points.";
+
+
+    @NotNull
+    public String getShortName() {
+        return "MultipleReturnPointsInspection";
+    }
+
+    @NotNull
+    public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, boolean isOnTheFly) {
+        return new PhpElementVisitor() {
+
+            public void visitPhpMethod(Method method) {
+                final PhpClass containingClass = method.getContainingClass();
+                final PsiElement nameIdentifier = method.getNameIdentifier();
+                if (containingClass != null && nameIdentifier != null && !containingClass.isTrait()) {
+
+                    final ReturnPointCountVisitor countVisitor = new ReturnPointCountVisitor(method);
+                    method.accept(countVisitor);
```

Comment:
`PsiTreeUtil` is good for traversing a tree upwards (to the root) or when you need to work with element's children. When you need to traverse a tree downwards (to the leafs) and you don't know an exact position of an element you're looking for or you need to work with several elements, a recursive element visitor is the best choice. It uses a "regular foreach loop" under the hood, but hides this irrelevant detail which makes the code more descriptive and allows concentrating on what you really want to do.

Working with PHP you usually use `PhpElementVisitor`. There's a recursive version of it called `PhpRecursiveElementVisitor`. It's marked as deprecated to recall you, that you should use it really carefully. When a tree is huge you can easily get a SOE or consume a lot of CPU for nothing. To avoid it a tree traversing should be stopped as soon as the desired element is found. For instance, the current implementation contains this mistake and will check each and every element for `instanceof PhpReturn`. A tree traversing won't be stopped even when more than 1 return point is found. That's why we usually don't recommend to use it at all.

**tl;dr**
1. Use `PsiTreeUtil` to find a parent or nearest children.
2. Use `PhpElementVisitor` to make your code more descriptive.
3. Never ever use `PhpRecursiveElementVisitor` (or any handmade recursive visitor) until you understand what you do. Don't forget to profile your code on really deep trees.